### ISSUE TITLE
P0: browser-budgeted real segmented WebGPU E2E（SmolLM2-135M first, then 1B scale-up）

### ISSUE THREAD

**Issue by azumag** [azumag/unzen#165]
## Background

LLM-proto の棚卸しで、核心である **実モデルの複数ブラウザ WebGPU 分散推論** がまだ未実証であることを確認した。

さらにP0設計を見直し、最初から 1B を2分割して各ブラウザへ700〜900MB級を持たせるのではなく、**通常ブラウザで現実的なartifactサイズを先に製品要件として固定**する。

## Browser artifact policy

1 browser worker あたり:

- target: **約200 MiB**
- preferred: **<=256 MiB**（P0はここを必須）
- normal: **<=512 MiB**
- degraded / exceptional: **<=1 GiB**
- hard reject: **>1 GiB**

1 GiB は上限であり目標ではない。初回download、Cache API/IndexedDB、WebGPU upload/compile、peak memory、短時間visitを考慮し、通常workerでは200MiB前後を狙う。

このpolicyは `src/browser-segment-artifact-budget.ts` と P0 preparation tool にコード化する。

## P0 model

最初の real E2E は **`onnx-community/SmolLM2-135M-ONNX` q4** を使う。

- LlamaForCausalLM
- 30 layers
- hidden size 576
- KV heads 3
- head size 64
- q4 source external weights 約182MB
- first split: layers 0..14 / 15..29

モデル品質ではなく、real transformerのlayer split / hidden-state relay / WebGPU continuationを証明するのが目的。

## Goal

```text
reference: full small model

split path:
Browser A
  first contiguous layer span
  -> real boundary tensors
  -> Coordinator-owned relay
Browser B
  remaining layers + output head
  -> logits/token
```

P0成功後、1B-classモデルは同じartifact policyを維持し、**segment数を増やして**scale-upする。大きいworkerにはSpanPipelineで複数の小artifactを連続assignしてrelay回数を減らす。

## Phase 1 — browser-budgeted split tooling

- SmolLM2-135M q4をP0 targetに固定するwrapperを追加
- layer 15 boundaryをgraphから検証してfail-fast
- segment-specific external weightsへrepack
- graph + weightsのbrowser artifact総byte数を計算
- P0では各segment <=256MiBを必須にする
- >256MiBなら大きいshardを許容せずsegment数を増やす方針でfail
- source / segment / external-data SHA-256をmanifestへ記録

## Phase 2 — same-machine real split correctness

- full model と segment0 -> boundary -> segment1 を同じtoken IDsで実行
- complete logits tolerance + top-1 token一致
- observed boundary byte sizeを記録
- KV-cache runtime hints: 3 heads / head size 64

## Phase 3 — real two-browser WebGPU relay

- Browser A / Browser B をdistinct worker identityとして接続
- Coordinator経由のみでboundary tensorsをrelay
- direct worker-to-workerは禁止
- segmentごとのweightsのみload
- SHA-256 verification + persistent Browser Cache API
- cold/warm artifact load / session create / execution timingをreport
- 1 token以上生成

## Phase 4 — failure / resume

- segment 0 checkpoint後にprimary Browser Bを切断/未実行
- standby Browser B' が同じCoordinator-owned checkpointを取得
- segment 0を再実行せずcontinuation

## Phase 5 — 1B scale-up

P0成立後に1B q4へ進む。

- target ~200MiB / segment
- preferred <=256MiB
- normal <=512MiB
- absolute <=1GiB
- graph weight-byte distributionから必要segment数を決める
- 2分割固定は禁止
- stable/high-capacity workerには複数隣接segmentをSpanPipelineでまとめてassign可能

## Stop boundary for automated implementation

コード、budget policy、split/repack tooling、fixture tests、Coordinator/browser harness、cache、docs、実行手順までは自動で進める。

**SmolLM2-135M q4実artifactを使う最初の real split run が必要になった地点で停止**し、再現コマンドと期待reportを提示する。

### Automated / CI
- [ ] browser artifact policyをコード化
- [ ] P0 wrapperがSmolLM2 geometryを固定
- [ ] P0 wrapperが>256MiB shardをfail
- [ ] deterministic split + boundary validation
- [ ] external-data repack / digest test
- [ ] full/split numerical fixture equivalence
- [ ] two-worker Coordinator relay test
- [ ] direct worker-to-worker rejection
- [ ] standby checkpoint reuse
- [ ] browser cache digest/cold-warm path
- [ ] typecheck / full test pass

### Real verification
- [ ] SmolLM2-135M q4を2 segmentへsplit
- [ ] 両segmentがpreferred <=256MiB
- [ ] same-machine real split inferenceがfull referenceと一致
- [ ] 2 distinct browsersで実WebGPU split inferenceを1 token以上完了
- [ ] observed boundary bytes / latency / cold-warm cacheを記録
- [ ] Browser B loss後にstandby resume

## Priority

**P0 / technical core blocker**。

本Issueのreal browser-budgeted E2Eが成立するまで、#158 Continuous Assurance production deploymentはHOLDのままとする。

## Related

- #33 2B / 2-worker simulated harness
- #39 checkpoint serialization / transfer measurement
- #101 simulated vs measured evidence separation
- #102 SegmentedModelManifest
- #103 Coordinator durability
- #158 Continuous Assurance production deployment（HOLD）
- PR #166
- `LLM-proto/docs/real-two-segment-webgpu-e2e.md`

**Comment by azumag** [comment]
## Automated implementation complete up to the real-verification boundary

Draft PR #166 now contains the automated portion of this P0 and latest CI is green on head `09952bfa73105e4b41500e0a69c261164ce0c87e`.

Implemented / verified before touching the real ~1.7 GB q4 artifact:

- deterministic layer 7/8 ONNX boundary discovery with fail-close drift detection
- dependency-closed segment 0 / segment 1 graph extraction
- streaming repack of only the external-data ranges used by each segment into independent `segment0.onnx_data` / `segment1.onnx_data`
- manifest binding source/segment/external-data SHA-256 + exact two-tensor boundary
- same-machine full-vs-split ONNX Runtime verifier with sequential session loading, complete-logits comparison and top-1 equality
- external-data ONNX fixture proving per-segment repack and exact full/split numerical equality
- localhost Coordinator relay for exactly two boundary tensors, no direct browser-to-browser route
- standby reuse of the existing Coordinator checkpoint without rerunning segment 0
- ONNX Runtime WebGPU browser harness with explicit `externalData` mapping for each segment
- Browser Cache API cold/warm artifact caching with SHA-256 verification and fail-close cache eviction on digest mismatch

Latest CI:
- typecheck ✅
- browser harness syntax checks ✅
- Python ONNX split/repack/numerical tests: **2/2** ✅
- full Vitest: **106 files / 749 tests** ✅
- real two-browser Coordinator contract: **3/3** ✅

### First point that now requires real environment confirmation

The next operation needs the actual local `onnx-community/Llama-3.2-1B-Instruct` q4 tree used in the prior WebGPU measurement.

Start with:

```bash
cd LLM-proto
python3 -m venv .venv-real-split
. .venv-real-split/bin/activate
pip install -r tools/requirements-real-split.txt
python3 tools/prepare_real_split.py \
  /absolute/path/to/onnx/model_q4.onnx \
  /absolute/path/to/unzen-split
```

The command must produce `segment0.onnx`, `segment0.onnx_data`, `segment1.onnx`, `segment1.onnx_data`, and `split-manifest.json`, with neither segment still referencing the original full external-data file.

After that, follow `LLM-proto/docs/real-two-segment-webgpu-e2e.md` for full-vs-split logits, two distinct localhost Chrome profiles, cold/warm cache, and standby resume.

#165 remains open until those real-model/browser checks pass. #158 remains HOLD.

**Comment by azumag** [comment]
## 2026-08-21 browser-budgeted P0 update

P0を1B/2-way splitから **SmolLM2-135M q4 / browser-budgeted split** へ切り替えた。

Automated implementation / CIは完了:

- browser artifact policy: target ~200MiB / preferred <=256MiB / normal <=512MiB / hard <=1GiB
- `prepare_browser_p0.py`: SmolLM2-135M (30 layers, split layer 15, hidden 576, KV heads 3) + preferred budget fail-close
- fixed revisionの実 `model_q4.onnx` graphをCIで取得・SHA-256固定してprobe
- 実graphのlayer 15 boundaryを確認:
  - `/model/layers.14/post_attention_layernorm/output_3`
  - `/model/layers.14/mlp/down_proj/MatMul/output_0`
  - いずれもlayer 14 producer、boundary tensor count=2、logits output=`logits`
- external-data split/repack fixture numerical gate: 4 Python tests pass
- browser artifact budget tests pass
- Coordinator relay / direct-network reject / standby reuse pass
- typecheck + browser syntax checks pass
- full Vitest: **107 files / 752 tests pass**

CI run: 32425207173, all steps success.

ここから先は real artifact verification。次は `onnx-community/SmolLM2-135M-ONNX` の q4 graph + 約182MB external weightsを実機へ置き、`tools/prepare_browser_p0.py` を実行して両segmentの実byte数が preferred <=256MiB になることを確認する。#165はreal browser E2Eまでopen、#158はHOLDのまま。

**Comment by azumag** [comment]
実機 P0 の結果を child handoff [#168](https://github.com/azumag/unzen/issues/168) に記録した。

- 固定 SmolLM2-135M q4 の split artifact は segment 0=`148,728,011` bytes、segment 1=`148,732,576` bytesで双方 preferred <=256 MiB
- same-machine full/split logits は shape 一致、maxAbsDiff `0.0`、top-1 `260` 一致
- Apple Metal 3 の非 fallback WebGPU で Browser A -> Coordinator -> Browser B の実 relay を PASS
- warm Cache API、Coordinator-only 403、standby checkpoint resume も PASS
- PR #166 head `31cbd3d` の custom-op checker fix と CI run `32427448089` も PASS

ただし今回の Chrome harness は2 controlled tabsで、strictなChrome profile identityを記録していないため、evidence levelは `self-reported-runtime` のまま。#158 は HOLD継続。

**Comment by azumag** [comment]
## 2026-08-21 P0 real verification status

#168 の実機reportを確認。technical coreは実測で成立:

- SmolLM2-135M q4 real split: PASS
- browser artifacts: segment0 148,728,011 bytes / segment1 148,732,576 bytes（双方 preferred <=256MiB）
- same-machine full/split logits: max abs diff 0.0 / top1 260一致
- real WebGPU segment0 -> Coordinator -> segment1: PASS
- boundary: 23,040 bytes
- Coordinator-only relay / direct route reject: PASS
- cold/warm Cache API: PASS
- standby resume without segment0 rerun: PASS

残件は strict Chrome profile isolation の機械証拠のみ。初回runはdistinct worker IDsだが同一profileの2 tabsを排除できなかった。

PR #166 head `462f3b7da75b6915d7c93b7e71da60841a6d9109` で Coordinator-issued HttpOnly profile probe を追加。同じChrome profileのworkersはresult POSTをHTTP 409 `profile-isolation-not-proven`で拒否し、distinct profileのみ `profileIsolationConfirmed=true` になる。

CI run `32428280204`: PASS（Python 5/5, Coordinator/profile 5/5, Vitest 107 files / 754 tests, pinned graph probe/typecheck/browser syntax PASS）。

#168 に最小rerun手順を追記済み。distinct Chrome profilesで `profileIsolationConfirmed=true` が取得できれば、本IssueのP0 acceptanceは全項目成立としてclose候補。#158はそれまではHOLD。
